nailgun/entity_mixins.py

```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+            #
+            attrs = {}
+            for field_name, field in fields.items():
+                if isinstance(field, OneToOneField):
+                    try:
+                        attrs[field_name] = _get_entity_id(field_name, result)
+                    except MissingValueError:
+                        pass
+                elif isinstance(field, OneToManyField):
+                    try:
+                        attrs[field_name] = _get_entity_ids(field_name, result)
+                    except MissingValueError:
+                        pass
+                else:
+                    try:
+                        attrs[field_name] = result[field_name]
+                    except KeyError:
+                        pass
+            normalized.append(attrs)
+        return normalized
+
+    def search(self, fields=None, query=None, filters=None):
+        """Search for entities.
+
+        At its simplest, this method searches for all entities of a given kind.
+        For example, to ask for all :class:`nailgun.entities.Organization`
+        entities::
+
+            Organization().search()
+
+        Values on an entity are used to generate a search query, and the
+        ``fields`` argument can be used to specify which fields should be used
+        when generating a search query::
+
+            org = Organization(name='foo', description='bar')
+            results = org.search()  # Search by name and description.
+            results = org.search(fields={'name', 'description'})  # Same.
+            results = org.search(fields={'name'})  # Search by name.
+            results = org.search(fields={'description'})  # Search by desc
+            results = org.search(fields=set())  # Search for all organizations.
+            results = org.search(fields={'title'})  # Error!
+
+        In some cases, the simple search queries that can be generated by
+        NailGun are not sufficient. In this case, you can pass in a raw search
+        query instead. For example, to search for all organizations with a name
+        of 'foo'::
+
+            Organization().search(query={'search': u'name="foo"'})
+
+        The example above is rather pointless: it is easier and more concise to
+        use a generated query. But — and this is a **very** important "but" —
+        the manual search query is melded in to the generated query. This can
+        be used to great effect::
+
+            Organization(name='foo').search(query={'per_page': 50})
+
+        In some cases, the server's search facilities may be insufficient, or
+        it may be inordinately difficult to craft a search query. In this case,
+        you can filter search results locally. For example, to ask the server
+        for a list of all organizations and then locally search through the
+        results for the organization named "foo"::
+
+            Organization().search(filters={'name': 'foo'})
```

Comment:
By the way, the more I think about an `expect` argument, the more I want to add one in. Consider how cool this would be:

```python
    organization = Organization().search(
        query={'search': 'name="Systems Engineering"'},
        expect=1,
    )[0]
    location = Location().search(
        query={'search': 'name="Boston"'},
        expect=1,
    )[0]
    hostgroup = HostGroup().search(
        query={'search': 'name="RHEV 3.5"'},
        expect=1,
    )[0]
```